File: e2e-tests/services/invoker-memory-kotlin-contracts/src/main/kotlin/dev/restate/sdktesting/invokermemory/MemoryPressureService.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate SDK Test suite tool,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-test-suite/blob/main/LICENSE
package dev.restate.sdktesting.invokermemory

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Name
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.kotlin.runBlock

@Service
@Name("MemoryPressureService")
class MemoryPressureService {
  @Handler
  suspend fun generate(input: String): String {
    for (i in 0 until 10) {
      runBlock { randomString(64.kb) }
    }
    return "ok-$input"
  }

  @Handler
  suspend fun generateOversized(input: String): String {
    // Single side effect producing 512KiB — exceeds the 256KiB per-invocation memory limit.
    // The invocation can never make progress and should be paused by the server.
    runBlock { randomString(512.kb) }
    return "ok-$input"
  }
}

File: e2e-tests/services/rust/src/memory_pressure_service.rs
// Copyright (c) 2023 - 2026 Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate e2e test suite,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/e2e/blob/main/LICENSE

//! `MemoryPressureService`, consumed by `InvokerMemoryTest` in the JVM e2e suite.
//!
//! The Kotlin test holds a matching `@Service` interface at
//! `e2e-tests/src/main/kotlin/dev/restate/sdktesting/contracts/MemoryPressureService.kt`;
//! the trait and handler names below must stay in sync with that contract.

use crate::util::{KB, random_string};
use restate_sdk::prelude::*;

#[restate_sdk::service]
#[name = "MemoryPressureService"]
pub trait MemoryPressureService {
    async fn generate(input: String) -> Result<String, HandlerError>;
    #[name = "generateOversized"]
    async fn generate_oversized(input: String) -> Result<String, HandlerError>;
}

pub struct MemoryPressureServiceImpl;

impl MemoryPressureService for MemoryPressureServiceImpl {
    async fn generate(&self, ctx: Context<'_>, input: String) -> Result<String, HandlerError> {
        for _ in 0..10 {
            ctx.run(|| async { Ok::<_, HandlerError>(random_string(64 * KB)) })
                .await?;
        }
        Ok(format!("ok-{input}"))
    }

    async fn generate_oversized(
        &self,
        ctx: Context<'_>,
        input: String,
    ) -> Result<String, HandlerError> {
        // 512 KiB single side effect — exceeds the 256 KiB per-invocation memory
        // limit so the runtime should pause the invocation.
        ctx.run(|| async { Ok::<_, HandlerError>(random_string(512 * KB)) })
            .await?;
        Ok(format!("ok-{input}"))
    }
}
